evaluator.go

```diff
@@ -104,26 +130,132 @@ func (state *State) SetDefaultCluster(name string) {
+	// splitting clusters in slices and spawn go routine for each
+	startTime := time.Now()
+	clusters := state.clusterNamesAsArray()
+	arrayOfClusterSlices := splitIntoSlices(clusters, 8)
+	var wg sync.WaitGroup
+	resultCh := make(chan Result, 8)
+	errorCh := make(chan error, 8)
+	defer close(resultCh)
+	defer close(errorCh)
+	for _, slice := range arrayOfClusterSlices {
+		wg.Add(1)
+		go func(s []string) {
+			defer wg.Done()
+			clusterLookupOnSlice(state, s, resultCh, errorCh)
+		}(slice)
+	}
+	done := make(chan interface{})
+	go func() {
+		wg.Wait()
+		close(done)
+	}()
+	results := mapset.NewSet()
 	errors := []error{}
 
-	// TODO: See if this is faster if parallelized (need to add coordination to
-	// cache).
-	for name, cluster := range state.clusters {
+Loop:
+	for {
+		select {
+		case r := <-resultCh:
+			results = results.Union(r.Set)
+		case err := <-errorCh:
+			errors = append(errors, err)
+		case <-done:
+			break Loop
+		}
+	}
+	// end of Loop:
+
+	wg.Wait()
+	state.metrics.buildTimeForClusterCache = time.Since(startTime).Seconds()
+	state.populateCachedCQRforSet(results)
+	state.metrics.buildTimeForCachedCQR =
+		time.Since(startTime).Seconds() - state.metrics.buildTimeForClusterCache
+	state.metrics.buildTimeForCache = time.Since(startTime).Seconds()
+	state.metrics.errorsDuringCacheBuild = len(errors)
+	state.metrics.cacheInitializedAt = time.Now()
+	return errors
+}
+
+// instead of returning entire state of metrics
+// returning map of metrics we want to expose.
+// should we return map[string] interface{} or map[string] int64 ?
+// decided to use int64, as it makes it easier on client side
+func (state *State) StateMetrics() map[string]int64 {
+	metrics := make(map[string]int64)
+	metrics["stateInitializedAt"] = state.metrics.initializedAt.Unix()
+	metrics["cacheInitializedAt"] = state.metrics.cacheInitializedAt.Unix()
+	metrics["numberOfClusters"] = int64(len(state.clusters))
+	metrics["numberOfCachedClusters"] = int64(state.clusterCache.Count())
+	metrics["numberOfcachedCQR"] = int64(state.cachedCQR.Count())
+	metrics["cacheTotalBuildTimeInSeconds"] = int64(state.metrics.buildTimeForCache)
+	metrics["cacheBuildTimeForClustersInSeconds"] = int64(state.metrics.buildTimeForClusterCache)
+	metrics["cacheBuildTimeForCQRInSeconds"] = int64(state.metrics.buildTimeForCachedCQR)
+	metrics["errorsDuringCacheBuild"] = int64(state.metrics.errorsDuringCacheBuild)
+	metrics["numberOfTruncatedResults"] = state.metrics.numberOfTruncatedResults
+	return metrics
+}
+
+func clusterLookupOnSlice(state *State, clusters []string,
+	resultCh chan<- Result, errorCh chan<- error) {
+	for _, clusterName := range clusters {
 		context := newContext()
-		context.currentClusterName = name
-		for key, _ := range cluster {
+		context.currentClusterName = clusterName
+		for key, _ := range state.clusters[clusterName] {
 			err := clusterLookup(state, &context, key)
+			// we are interested only results for key CLUSTER
```

Comment:
I think it's because we're doing two things at once: priming the cache for all keys, but also fetching out the CLUSTER results to subsequently pass to populate CQR.

maybe `primeCacheForCluster` or something like that is a more appropriate name, with function documentation explaining that?

Also is there any reason to thread the channels in here? I feel like it needlessly couples - we could just instead return the cluster or an error and have the calling function put them in to the channels?